src/v/serde/rw/variant.h

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/vlog.h"
+#include "serde/rw/rw.h"
+
+#include <array>
+#include <type_traits>
+#include <utility>
+#include <variant>
+
+namespace serde {
+
+// A small wrapper around std::variant that is marked to be serializable.
+//
+// Special precation needs to be taken to mark a variant type as serializable
+// with respect to compatibility. `serde::variant` should be a drop in
+// replacement for `std::variant`, but allows for serde operations.
+//
+// # Variant Wire Compatibility:
+//
+// Variants are tricky w.r.t. compatibility, they are forward compatible if new
+// discrimiants are appended to the variant.
+//
+// Concretely that means that you can only read old versions of a variant if new
+// options are appended. Writing a new variant option should happen with a
+// feature barrier, so that an older reader will never read a newly written
+// variant.
```

Comment:
this is what compat versioned envelopes are for. should serde::variant wrapper also be an envelope?

Reply:
I am having a hard time seeing how that would work. In my PR where I am using this I wrap the variant in an enveloped struct. I can recommend that here?

I can also drop this change and manually implement the variant for my structure, maybe thats better?

Reply:
> I can also drop this change and manually implement the variant for my structure, maybe thats better?

I don't think we need to drop it, but we want to make sure it works going forward because it's a forever decision?

> In my PR where I am using this I wrap the variant in an enveloped struct.

```c++
struct transform_offset_options
  : serde::envelope<
      transform_offset_options,
      serde::version<0>,
      serde::compat_version<0>> {

    serde::variant<latest_offset, model::timestamp> position;
    auto serde_fields() { return std::tie(position); }
```

This isn't the same thing as `position` itself being a versioned envelope. Imagine we tried to support std::variant in a non-generic way for some situation with an envelope:

```c++
struct thing : env<thing, version<0>, compat<0>> {
  enum value_type { a, b };
  value_type t;
  value_a_type a;
  value_b_type b;

  write() { write(t); switch(t) { write(a) or write(b); } }
  read() { t = read_int(); switch(t) { read(a) or read(b); } }
};
```

This is _equivalent_ in terms of information content to what's happening with std::variant in this PR (index + bytes for the held variant value), except, it also has a version, compat version, and serde encodes the size of the entire envelope on the wire.

Does this formulation offer any fundamental advantage? If yes, then it seems like serde::variant would also miss that advantage.

One thing that comes to mind is that with `thing` is that `version` kind of exists implicitly on either side of the wire in the form of std::variant::size. Appending a new type to the variant is like bumping the version, but we don't communicate it. Should we?

node0
std::variant<a> v1;

node1
std::variant<a,b> v2;

if node0 sent a@v1 to node1 then on the write it would be like 1:0:<data> where 1 is the size of the variant and 0 is the index of the held alternative. node1 would observe that 1<=2 and be happy--this is like the version check in serde today.

in serde if we append data to an envelope, an older version can still decode this by skipping unknown data (this is where the envelope side comes into play). appending a new variant type is not the same thing. that is more analogous to just changing the type of a an existing field, which can't be handled automatically--either a custom reader needs to be defined, or compat version needs to be bumped.

but in the example above, if node1 sends a@v2 to node0 it would be encoded as 2:0:<data>. Today that would be 0:<data> and that node0 would be happy because `index < std::variant_size_v<UnderlyingType>` , and, it actually couldn't be unhappy because there isn't enough information. But I think I'm arguing that it should be unhappy because the type of the field changed.

So I'm wondering, if the tag_invoke for read should actually be

```
    auto index = read_nested<size_t>(in, bytes_left_limit);
    auto size = read_nested<size_t>(in, bytes_left_limit);

    if (size != std::variant_size_v<UnderlyingType>) { like a compat violation }
    if (index >= std::variant_size_v<UnderlyingType>) { same }
```

It just seems like std::variant encoding here _wants_ to be like an envelope, but just isn't quite going all the way.

Reply:
> But I think I'm arguing that it should be unhappy because the type of the field changed.
`size != std::variant_size_v<UnderlyingType>`

This seems overly strict right? Like you mentioned in appending fields is fine in envelope, and should be in serde::variant IMO. It's fine because the variant is still the same value. I see the argument for making it a type change, but I guess it's if you see variant as an atomic or composite value. I argue it's a composite value, adding a field is the same. I view the compat story here is the same as a protobuf `oneof` or serde::envelope, which it's fine to append new types without bumping versions. If you need to bump versions (make incompatible changes) you'll need to bump the version on the parent `envelope` and handle parsing the old variant. I guess you're arguing for needing to do that either always on the parent type or never on the parent type (adding versions to serde::variant). I see the consistency but it is the least convenient, and is less consistent when you compare it to serde::envelope. Because serde::variant won't have the ability to have custom decode/encode functions like a custom struct I do think it should always be wrapped in an envelope. Since backwards compatibility issues are extremely hard to reason about and test, I would argue the safest defaults make sense.

I think the safest default is to add serde::version to force people to think about this? Then you only have to bump the version if you make an incompatible change. Thoughts @dotnwat

Reply:
> size != std::variant_size_v<UnderlyingType>
> This seems overly strict right?

yeh, it's one end of the spectrum, i agree. i think the thing we should guard against is trying to catch mistakes.

In the coverletter:

> Writing a new variant option should happen with a feature barrier

it would be nice if a serde reader is always able to detect the backwards incompatible change and throw. using the feature barrier is then the way that we arrange for the system to not to throw, but the safe guard is present. today we do that by bumping compact_version, and crucially, changing compat_version is lexicographically close to type that is undergoing the incompatible change.

> bump the version on the parent envelope

So if a serde_variant is encoded as part of multiple envelopes, then one has to go update all the envelopes.

Anyway, I think we can merge this. It strikes me as dangerous and should have a compat version embedded--it's just a byte, but I also don't think there will be very many users so 🤷

Reply:
> i think the thing we should guard against is trying to catch mistakes.

That's fine, I would probably err on the side of being flexible, because normally these things are only caught too late, but I've updated the PR to add the size, better to be noisy and throw than be silent and have UB